sdks/java/core/src/main/java/org/apache/beam/sdk/testing/BigqueryMatcher.java

```diff
@@ -151,17 +151,24 @@ QueryResponse queryWithRetries(Bigquery bigqueryClient, QueryRequest queryConten
     IOException lastException = null;
     do {
       try {
-        return bigqueryClient.jobs().query(projectId, queryContent).execute();
+        response = bigqueryClient.jobs().query(projectId, queryContent).execute();
+        if (response != null && response.getJobComplete()) {
+          return response;
+        }
       } catch (IOException e) {
         // ignore and retry
         LOG.warn("Ignore the error and retry the query.");
         lastException = e;
       }
     } while(BackOffUtils.next(sleeper, backOff));
-    throw new IOException(
-        String.format(
-            "Unable to get BigQuery response after retrying %d times", MAX_QUERY_RETRIES),
-        lastException);
+
+    if (lastException != null) {
+      throw new IOException(
+          String.format(
+              "Unable to get BigQuery response after retrying %d times", MAX_QUERY_RETRIES),
+          lastException);
+    }
+    return response;
```

Comment:
cool.

The reason I want to catch IOException inside and re-throw it mainly because I want to keep original error info and stack trace for user debugging.